client/lib/happychat/connection-ng.js

```diff
@@ -0,0 +1,172 @@
+							dispatch( receiveInit( { signer_user_id, locale, groups, geoLocation } ) );
+							dispatch( requestTranscript() );
+							resolve( socket );
+						} )
+						.on( 'unauthorized', () => {
+							socket.close();
+							dispatch( receiveUnauthorized( 'User is not authorized' ) );
+							reject( 'User is not authorized' );
+						} )
+						.on( 'disconnect', reason => dispatch( receiveDisconnect( reason ) ) )
+						.on( 'reconnecting', () => dispatch( receiveReconnecting() ) )
+						.on( 'status', status => dispatch( receiveStatus( status ) ) )
+						.on( 'accept', accept => dispatch( receiveAccept( accept ) ) )
+						.on( 'message', message => dispatch( receiveMessage( message ) ) );
+				} )
+				.catch( e => reject( e ) );
+		} );
+
+		return this.openSocket;
+	}
+
+	/**
+	 * Given a Redux action, emits a SocketIO event.
+	 *
+	 * @param  { Object } action A Redux action with props
+	 *                    {
+	 *                  		event: SocketIO event name,
+	 *                  	  payload: contents to be sent,
+	 *                  	  error: message to be shown should the event fails to be sent,
+	 *                  	}
+	 * @return { Promise } Fulfilled (returns nothing)
+	 *                     or rejected (returns an error message)
+	 */
+	send( action ) {
+		if ( ! this.openSocket ) {
+			return;
+		}
+		return this.openSocket.then(
+			socket => socket.emit( action.event, action.payload ),
+			e => {
+				this.dispatch( receiveError( 'failed to send ' + action.event + ': ' + e ) );
+				// so we can relay the error message, for testing purposes
+				return Promise.reject( e );
+			}
+		);
+	}
+
+	/**
+	 *
+	 * Given a Redux action and a timeout, emits a SocketIO event that request
+	 * some info to the Happychat server.
+	 *
+	 * The request can have three states, and will dispatch an action accordingly:
+	 *
+	 * - request was succesful: would dispatch action.callback
+	 * - request was unsucessful: would dispatch receiveError
+	 * - request timeout: would dispatch action.callbackTimeout
+	 *
+	 * @param  { Object } action A Redux action with props
+	 *                  	{
+	 *                  		event: SocketIO event name,
+	 *                  		payload: contents to be sent,
+	 *                  		callback: a Redux action creator,
+	 *                  		callbackTimeout: a Redux action creator,
+	 *                  	}
+	 * @param  { Number } timeout How long (in milliseconds) has the server to respond
+	 * @return { Promise } Fulfilled (returns the transcript response)
+	 *                     or rejected (returns an error message)
+	 */
+	request( action, timeout ) {
+		if ( ! this.openSocket ) {
+			return;
+		}
+
+		return this.openSocket.then(
+			socket => {
+				const promiseRace = Promise.race( [
+					new Promise( ( resolve, reject ) => {
+						socket.emit( action.event, action.payload, ( e, result ) => {
+							if ( e ) {
+								return reject( new Error( e ) ); // request successful
```

Comment:
I think the comments need to be swaped.

Reply:
That's true! I already fixed this in an upcoming PR that merge this in the real `connection.js` code. Going to merge this as it is as to avoid integration issue.